aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteModelTree.java

```diff
@@ -0,0 +1,157 @@
+
+package com.amplifyframework.datastore.storage.sqlite;
+
+import android.database.Cursor;
+import android.database.sqlite.SQLiteDatabase;
+import androidx.annotation.NonNull;
+
+import com.amplifyframework.core.model.Model;
+import com.amplifyframework.core.model.ModelAssociation;
+import com.amplifyframework.core.model.ModelSchema;
+import com.amplifyframework.core.model.ModelSchemaRegistry;
+import com.amplifyframework.core.model.query.QueryOptions;
+import com.amplifyframework.core.model.query.Where;
+import com.amplifyframework.core.model.query.predicate.QueryField;
+import com.amplifyframework.core.model.query.predicate.QueryPredicate;
+import com.amplifyframework.core.model.query.predicate.QueryPredicateOperation;
+import com.amplifyframework.core.model.query.predicate.QueryPredicates;
+import com.amplifyframework.datastore.DataStoreException;
+import com.amplifyframework.datastore.storage.sqlite.adapter.SQLiteTable;
+import com.amplifyframework.util.Empty;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utility class to help traverse a tree of models by relationship.
+ */
+final class SQLiteModelTree {
+    private final ModelSchemaRegistry registry;
+    private final SQLCommandFactory commandFactory;
+    private final SQLiteDatabase database;
+
+    /**
+     * Constructs a model family tree traversing utility.
+     * @param registry model registry to search schema from
+     * @param commandFactory SQL command factory
+     * @param database SQLite database connection handle
+     */
+    SQLiteModelTree(ModelSchemaRegistry registry,
+                    SQLCommandFactory commandFactory,
+                    SQLiteDatabase database) {
+        this.registry = registry;
+        this.commandFactory = commandFactory;
+        this.database = database;
+    }
+
+    /**
+     * Returns a map of descendants of a set of models (of same type).
+     * A model is a child of its parent if it uses its parent's ID as foreign key.
+     * @param root Collection of models to query its descendants of.
+     * @return Map of descendants keyed by model schema. The value contains a set of
+     *          descendants' IDs for that model type.
+     */
+    <T extends Model> Map<ModelSchema, Set<String>> descendantsOf(Collection<T> root) {
+        if (Empty.check(root)) {
+            throw new IllegalArgumentException("Cannot traverse tree from an empty root.");
+        }
+        Map<ModelSchema, Set<String>> descendants = new LinkedHashMap<>();
+        ModelSchema rootSchema = registry.getModelSchemaForModelInstance(root.iterator().next());
+        Set<String> rootIds = new HashSet<>();
+        for (T model : root) {
+            rootIds.add(model.getId());
+        }
+        recurseTree(descendants, rootSchema, rootIds);
+        return descendants;
+    }
+
+    private void recurseTree(
+            Map<ModelSchema, Set<String>> tree,
```

Comment:
This is just a `Map`, not a tree, so maybe call it `map`?.

```suggestion
            Map<ModelSchema, Set<String>> map,
```